src/Compilers/CSharp/Portable/Binder/PatternSwitchBinder2.cs

```diff
@@ -0,0 +1,234 @@
+            var locals = GetDeclaredLocalsForScope(node);
+            var functions = GetDeclaredLocalFunctionsForScope(node);
+            BoundDecisionDag decisionDag = null; // we'll compute it later
+            return new BoundPatternSwitchStatement(
+                syntax: node,
+                expression: boundSwitchExpression,
+                someLabelAlwaysMatches: someCaseMatches,
+                innerLocals: locals,
+                innerLocalFunctions: functions,
+                switchSections: switchSections,
+                defaultLabel: defaultLabel,
+                breakLabel: this.BreakLabel,
+                decisionDag: decisionDag,
+                isComplete: isComplete);
+        }
+
+        /// <summary>
+        /// Bind a pattern switch label in order to force inference of the type of pattern variables.
+        /// </summary>
+        internal override void BindPatternSwitchLabelForInference(CasePatternSwitchLabelSyntax node, DiagnosticBag diagnostics)
+        {
+            // node should be a label of this switch statement.
+            Debug.Assert(this.SwitchSyntax == node.Parent.Parent);
+
+            // This simulates enough of the normal binding path of a switch statement to cause
+            // the label's pattern variables to have their types inferred, if necessary.
+            // It also binds the when clause, and therefore any pattern and out variables there.
+            BoundPatternSwitchLabel defaultLabel = null;
+            BindPatternSwitchSectionLabel(
+                sectionBinder: GetBinder(node.Parent),
+                node: node,
+                label: LabelsByNode[node],
+                defaultLabel: ref defaultLabel,
+                diagnostics: diagnostics);
+        }
+
+        /// <summary>
+        /// Bind the pattern switch labels, reporting in the process which cases are subsumed. The strategy
+        /// implemented with the help of <see cref="SubsumptionDiagnosticBuilder"/>, is to start with an empty
+        /// decision tree, and for each case we visit the decision tree to see if the case is subsumed. If it
+        /// is, we report an error. If it is not subsumed and there is no guard expression, we then add it to
+        /// the decision tree.
+        /// </summary>
+        private ImmutableArray<BoundPatternSwitchSection> BindPatternSwitchSections(
+            Binder originalBinder,
+            out BoundPatternSwitchLabel defaultLabel,
+            out bool isComplete,
+            out bool someCaseMatches,
+            DiagnosticBag diagnostics)
+        {
+            defaultLabel = null;
+
+            // someCaseMatches will be set to true if some single case label would handle all inputs
+            someCaseMatches = false;
+
+            // Bind match sections
+            var boundPatternSwitchSectionsBuilder = ArrayBuilder<BoundPatternSwitchSection>.GetInstance();
```

Comment:
>GetInstance [](start = 92, length = 11)

Do we know the required size for this builder?

Reply:
Resolved in [https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec](https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec) of [https://github.com/dotnet/roslyn/pull/23357](https://github.com/dotnet/roslyn/pull/23357)

---
In reply to: [156525269](https://github.com/dotnet/roslyn/pull/23209#discussion_r156525269) [](ancestors = 156525269)